File: src/nodejs/src/services/PatientService.ts
import "../mappers/PatientMapper";
import { ApiDataResult } from "../dtos/ApiResult";
import { BaseService } from "./BaseService";
import { CreatePatientRequest } from "../dtos/CreatePatientRequest";
import { IPatientRepository } from "../repositories/IPatientRepository";
import { IPatientService } from "./IPatientService";
import { Order } from "../entities/Order";
import { PagedRequest } from "../dtos/PagedRequest";
import { PagedResult } from "../dtos/PagedResult";
import { Patient } from "../entities/Patient";
import { PatientDto } from "../dtos/PatientDto";
import { toDtos } from "../mappers/PatientMapper";
import { DateUtils } from "../utils/dateUtils";

/**
 * Service for managing patient-related business logic.
 */
export class PatientService extends BaseService implements IPatientService {
  /**
   * Creates a new instance of PatientService.
   * @param patientRepository - The patient repository.
   */
  constructor(private patientRepository: IPatientRepository) {
    super();
  }

  /**
   * Retrieves a paginated list of patients within a date range.
   * @param request - The pagination request.
   * @param startTime - Optional start of the date range.
   * @param endTime - Optional end of the date range.
   * @returns An API result containing the paginated patient list.
   */
  async getPatientsAsync(
    request: PagedRequest,
    startTime?: Date,
    endTime?: Date
  ): Promise<ApiDataResult<PagedResult<PatientDto>>> {
    // Use UTC epoch as default start, current UTC time as default end
    const start = startTime || new Date(0);
    const end = endTime || DateUtils.utcNow();

    if (start > end) {
      return super.badRequestDefaultDataResult<PagedResult<PatientDto>>(
        "The start date must be earlier than or equal to the end date."
      );
    }

    const threeYearsLimit = 3 * 365 + 1;
    const durationMs = end.getTime() - start.getTime();
    const durationDays = Math.floor(durationMs / (1000 * 60 * 60 * 24));

    if (durationDays > threeYearsLimit) {
      return super.badRequestDefaultDataResult<PagedResult<PatientDto>>(
        `The date range must not exceed 3 years. Requested duration was ${durationDays} days.`
      );
    }

    const [patients, totalCount] = await this.patientRepository.getPatientsAsync(
      start,
      end,
      request.pageNumber,
      request.pageSize
    );

    if (totalCount === 0) {
      return super.noDataFoundPagedResult<PatientDto>("No patients found for the given date range.");
    }

    const patientDtos = toDtos(patients);
    return super.successPagedResult(patientDtos, totalCount, request.pageNumber, request.pageSize);
  }

  /**
   * Retrieves a patient by ID.
   * @param id - The patient ID.
   * @returns An API result containing the patient if found, null otherwise.
   */
  async getPatientAsync(id: number): Promise<ApiDataResult<PatientDto | null>> {
    const patient = await this.patientRepository.getPatientAsync(id);

    if (!patient) {
      return super.noDataFoundDataResult<PatientDto | null>(null, "Patient not found");
    }

    return super.successDataResult(patient.toDto());
  }

  /**
   * Creates a new patient.
   * @param request - The create patient request.
   * @returns An API result containing the created patient.
   */
  async createPatientAsync(request: CreatePatientRequest): Promise<ApiDataResult<PatientDto>> {
    const newOrder = new Order();
    newOrder.message = request.orderMessage;

    const newPatient = new Patient();
    newPatient.name = request.name;
    newPatient.orders = [newOrder];

    const savedPatient = await this.patientRepository.addAsync(newPatient);
    return super.successDataResult(savedPatient.toDto(), "Patient created");
  }
}

File: src/python/app/services/patient_service.py
"""Patient service."""

import math
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.entities import Patient, Order
from app.repositories import PatientRepository
from app.schemas import PatientDto, GetPatientsRequest, CreatePatientRequest, ApiResultData, PagedResult, ApiCode


class PatientService:
    """Service for patient business logic."""

    def __init__(self, db: AsyncSession):
        """Initialize service with database session."""
        self.db = db
        self.patient_repo = PatientRepository(db)

    async def get_patients(self, request: GetPatientsRequest) -> ApiResultData[PagedResult[PatientDto]]:
        """
        Get patients within date range with pagination.

        Args:
            request: Request with date range and pagination parameters

        Returns:
            API result with paged patient data
        """
        try:
            patients, total_count = await self.patient_repo.get_patients(
                start_time=request.start_time,
                end_time=request.end_time,
                page_number=request.page_number,
                page_size=request.page_size,
            )

            total_pages = math.ceil(total_count / request.page_size) if total_count > 0 else 1
            patient_dtos = [PatientDto.model_validate(patient) for patient in patients]
            paged_result = PagedResult[PatientDto](
                data=patient_dtos,
                page_number=request.page_number,
                total_pages=total_pages,
                total_count=total_count,
                page_size=request.page_size,
            )

            return ApiResultData[PagedResult[PatientDto]](
                success=True, code=ApiCode.SUCCESS, message="Patients retrieved successfully", data=paged_result
            )

        except Exception as e:
            return ApiResultData[PagedResult[PatientDto]](
                success=False,
                code=ApiCode.DATA_ACCESS_ERROR,
                message=f"Failed to retrieve patients: {str(e)}",
                data=None,
            )

    async def get_patient(self, patient_id: int) -> ApiResultData[PatientDto]:
        """
        Get patient by ID.

        Args:
            patient_id: Patient ID

        Returns:
            API result with patient data
        """
        try:
            patient = await self.patient_repo.get_patient(patient_id)

            if not patient:
                return ApiResultData[PatientDto](
                    success=False,
                    code=ApiCode.NO_DATA_FOUND,
                    message=f"Patient with ID {patient_id} not found",
                    data=None,
                )

            patient_dto = PatientDto.model_validate(patient)
            return ApiResultData[PatientDto](
                success=True, code=ApiCode.SUCCESS, message="Patient retrieved successfully", data=patient_dto
            )

        except Exception as e:
            return ApiResultData[PatientDto](
                success=False,
                code=ApiCode.DATA_ACCESS_ERROR,
                message=f"Failed to retrieve patient: {str(e)}",
                data=None,
            )

    async def create_patient(self, request: CreatePatientRequest) -> ApiResultData[PatientDto]:
        """
        Create a new patient with an initial order.

        Args:
            request: Request with patient and order data

        Returns:
            API result with created patient data
        """
        try:
            patient = Patient(name=request.name)
            order = Order(message=request.order_message)
            patient.orders.append(order)

            created_patient = await self.patient_repo.add(patient)
            await self.db.commit()
            await self.db.refresh(created_patient, ["orders"])

            patient_dto = PatientDto.model_validate(created_patient)
            return ApiResultData[PatientDto](
                success=True, code=ApiCode.SUCCESS, message="Patient created successfully", data=patient_dto
            )

        except Exception as e:
            await self.db.rollback()
            return ApiResultData[PatientDto](
                success=False, code=ApiCode.OPERATION_FAILED, message=f"Failed to create patient: {str(e)}", data=None
            )
